pkg/server/container_execsync.go

```diff
@@ -48,92 +43,68 @@ func (c *criContainerdService) ExecSync(ctx context.Context, r *runtime.ExecSync
+	if err != nil {
+		return nil, fmt.Errorf("failed to get container spec: %v", err)
 	}
-	pspec := spec.Process
-	pspec.Args = r.GetCmd()
-	rawSpec, err := json.Marshal(pspec)
+	task, err := container.Task(ctx, nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to marshal oci process spec %+v: %v", pspec, err)
+		return nil, fmt.Errorf("failed to load task: %v", err)
 	}
 
-	// TODO(random-liu): Replace the following logic with containerd client and add unit test.
-	// Prepare streaming pipes.
-	execDir, err := ioutil.TempDir(getContainerRootDir(c.rootDir, id), "exec")
+	pspec := spec.Process
+	pspec.Args = r.GetCmd()
+
+	execID := generateID()
+	stdinBuf, stdoutBuf, stderrBuf := new(bytes.Buffer), new(bytes.Buffer), new(bytes.Buffer)
+	io := containerd.NewIOWithTerminal(stdinBuf, stdoutBuf, stderrBuf, pspec.Terminal)
+	process, err := task.Exec(ctx, execID, pspec, io)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create exec streaming directory: %v", err)
+		return nil, fmt.Errorf("failed to create exec %q: %v", execID, err)
 	}
 	defer func() {
-		if err = c.os.RemoveAll(execDir); err != nil {
-			glog.Errorf("Failed to remove exec streaming directory %q: %v", execDir, err)
+		if _, err := process.Delete(ctx); err != nil {
+			glog.Errorf("Failed to delete exec process %q for container %q: %v", execID, id, err)
 		}
 	}()
-	_, stdout, stderr := getStreamingPipes(execDir)
-	_, stdoutPipe, stderrPipe, err := c.prepareStreamingPipes(ctx, "", stdout, stderr)
-	if err != nil {
-		return nil, fmt.Errorf("failed to prepare streaming pipes: %v", err)
-	}
-	defer stdoutPipe.Close()
-	defer stderrPipe.Close()
-
-	// Start redirecting exec output.
-	stdoutBuf, stderrBuf := new(bytes.Buffer), new(bytes.Buffer)
-	go io.Copy(stdoutBuf, stdoutPipe) // nolint: errcheck
-	go io.Copy(stderrBuf, stderrPipe) // nolint: errcheck
 
 	// Get containerd event client first, so that we won't miss any events.
 	// TODO(random-liu): Add filter to only subscribe events of the exec process.
+	// TODO(random-liu): Use `Wait` after is fixed. (containerd#1279, containerd#1287)
 	cancellable, cancel := context.WithCancel(ctx)
 	eventstream, err := c.eventService.Subscribe(cancellable, &events.SubscribeRequest{})
 	if err != nil {
-		return nil, fmt.Errorf("failed to get containerd event: %v", err)
+		return nil, fmt.Errorf("failed to subscribe event stream: %v", err)
 	}
 	defer cancel()
 
-	execID := generateID()
-	_, err = c.taskService.Exec(ctx, &tasks.ExecProcessRequest{
-		ContainerID: id,
-		Terminal:    false,
-		Stdout:      stdout,
-		Stderr:      stderr,
-		Spec: &prototypes.Any{
-			TypeUrl: runtimespec.Version,
-			Value:   rawSpec,
-		},
-		ExecID: execID,
-	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to exec in container %q: %v", id, err)
+	if err := process.Start(ctx); err != nil {
+		return nil, fmt.Errorf("failed to start exec %q: %v", execID, err)
 	}
+
 	exitCode, err := c.waitContainerExec(eventstream, id, execID)
```

Comment:
Let's use `Wait` after we bump up containerd version including the fixes for `Wait`.